intermediate_source/rpc_param_server_tutorial.rst

```diff
@@ -0,0 +1,398 @@
+
+           x = self.dropout1(x)
+           x = torch.flatten(x, 1)
+           # Move tensor to next device if necessary
+           next_device = next(self.fc1.parameters()).device
+           x = x.to(next_device)
+
+           x = self.fc1(x)
+           x = F.relu(x)
+           x = self.dropout2(x)
+           x = self.fc2(x)
+           output = F.log_softmax(x, dim=1)
+           return output
+
+Next, let's define some helper functions that will be useful for the rest of our script. The following uses `rpc_sync <https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.rpc_sync>`_ and `RRef <https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.RRef>`_ in order to define a function that invokes a given method on an object living on a remote node. Below, our handle to the remote object is given by the ``rref`` argument, and we run it on its owning node: ``rref.owner()``. On the caller node, we run this command synchronously through the use of ``rpc_sync``\ , meaning that we will block until a response is received.
+
+.. code-block:: python
+
+   # --------- Helper Methods --------------------
+
+   # On the local node, call a method with first arg as the value held by the
+   # RRef. Other args are passed in as arguments to the function called.
+   # Useful for calling instance methods. method could be any matching function, including 
+   # class methods.
+   def call_method(method, rref, *args, **kwargs):
+       return method(rref.local_value(), *args, **kwargs)
+
+   # Given an RRef, return the result of calling the passed in method on the value
+   # held by the RRef. This call is done on the remote node that owns
+   # the RRef and passes along the given argument.
+   # Example: If the value held by the RRef is of type Foo, then
+   # remote_method(Foo.bar, rref, arg1, arg2) is equivalent to calling
+   # <foo_instance>.bar(arg1, arg2) on the remote node and getting the result
+   # back.
+
+   def remote_method(method, rref, *args, **kwargs):
+       args = [method, rref] + list(args)
+       return rpc.rpc_sync(rref.owner(), call_method, args=args, kwargs=kwargs)
+
+Now, we're ready to define our parameter server. We will subclass ``nn.Module`` and save a handle to our network defined above. We'll also save an input device which will be the device our input is transferred to before invoking the model.
+
+.. code-block:: python
+
+   # --------- Parameter Server --------------------
+   class ParameterServer(nn.Module):
+       def __init__(self, num_gpus=0):
+           super().__init__()
+           model = Net(num_gpus=num_gpus)
+           self.model = model
+           self.input_device = torch.device(
+               "cuda:0" if torch.cuda.is_available() and num_gpus > 0 else "cpu")
+
+Next, we'll define our forward pass. Note that regardless of the device of the model output, we move the output to CPU, as the Distributed RPC Framework currently only supports sending CPU tensors over RPC.
```

Comment:
Let's say we intentionally disabled sending CUDA tensors to avoid errors when tensor device is not available on the callee.

Reply:
We disable it in all cases, right? This would sort of imply we conditionally disable it based on what devices are available on the calleee. How about "We have intentionally disabled sending CUDA tensors over RPC due to the potential for different devices (CPU/GPU) on on the caller/callee"